### ISSUE TITLE
`tracing-appender = "0.2.2"` code on crates.io does not match with the code in github!

### ISSUE THREAD

**Issue by ozgunozerk** [tokio-rs/tracing#2168]
### Platform
macos (but I think its not related to OS)

### Crates

appender

### Description
TL;DR -> the code in repo does not match against crates.io release
for the appender, when I use: `tracing-appender = "0.2.2"` in my **Cargo.toml** it works.
But when I use `tracing-appender = { git = "https://github.com/tokio-rs/tracing", branch = "v0.1.x" }, or any of the revisions related to 0.2.2, it does not work. 

<details>
    <summary> It does not work explanation: </summary>

```
error[E0277]: expected a `Fn<()>` closure, found `NonBlocking`
  --> src/main.rs:49:71
   |
49 |             BunyanFormattingLayer::new("subspace-desktop".to_owned(), non_blocking)
   |             --------------------------                                ^^^^^^^^^^^^ expected an `Fn<()>` closure, found `NonBlocking`
   |             |
   |             required by a bound introduced by this call
   |
   = help: the trait `Fn<()>` is not implemented for `NonBlocking`
   = note: wrap the `NonBlocking` in a closure with no arguments: `|| { /* code */ }`
   = note: required because of the requirements on the impl of `for<'a> MakeWriter<'a>` for `NonBlocking`
note: required by a bound in `BunyanFormattingLayer::<W>::new`
  --> /Users/ozgunozerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-bunyan-formatter-0.3.2/src/formatting_layer.rs:53:9
   |
53 | impl<W: for<'a> MakeWriter<'a> + 'static> BunyanFormattingLayer<W> {
   |         ^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `BunyanFormattingLayer::<W>::new`

error[E0599]: the method `and_then` exists for struct `BunyanFormattingLayer<NonBlocking>`, but its trait bounds were not satisfied
   --> src/main.rs:50:18
    |
50  |                 .and_then(JsonStorageLayer)
    |                  ^^^^^^^^ method cannot be called on `BunyanFormattingLayer<NonBlocking>` due to unsatisfied trait bounds
    |
   ::: /Users/ozgunozerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-bunyan-formatter-0.3.2/src/formatting_layer.rs:44:1
    |
44  | pub struct BunyanFormattingLayer<W: for<'a> MakeWriter<'a> + 'static> {
    | --------------------------------------------------------------------- doesn't satisfy `_: tracing_subscriber::Layer<_>`
    |
   ::: /Users/ozgunozerk/.cargo/git/checkouts/tracing-e9bbb56ea31f0c18/68bc91f/tracing-appender/src/non_blocking.rs:126:1
    |
126 | pub struct NonBlocking {
    | ---------------------- doesn't satisfy `NonBlocking: MakeWriter<'a>`
    |
    = note: the following trait bounds were not satisfied:
            `NonBlocking: MakeWriter<'a>`
            which is required by `BunyanFormattingLayer<NonBlocking>: tracing_subscriber::Layer<_>`

error[E0277]: expected a `Fn<()>` closure, found `NonBlocking`
  --> src/main.rs:49:13
   |
49 |             BunyanFormattingLayer::new("subspace-desktop".to_owned(), non_blocking)
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected an `Fn<()>` closure, found `NonBlocking`
   |
   = help: the trait `Fn<()>` is not implemented for `NonBlocking`
   = note: wrap the `NonBlocking` in a closure with no arguments: `|| { /* code */ }`
   = note: required because of the requirements on the impl of `for<'a> MakeWriter<'a>` for `NonBlocking`
note: required by a bound in `BunyanFormattingLayer`
  --> /Users/ozgunozerk/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-bunyan-formatter-0.3.2/src/formatting_layer.rs:44:37
   |
44 | pub struct BunyanFormattingLayer<W: for<'a> MakeWriter<'a> + 'static> {
   |                                     ^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `BunyanFormattingLayer`

Some errors have detailed explanations: E0277, E0599.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `subspace-desktop` due to 3 previous errors
```

</details>

---

Long story:
I'm using tracing-appender in my application along with `tracing-bunyan-formatter = "0.3.2"`.
They are working fine. But there was a feature I wanted to implement to `appender`. And I did it, it is working (want to open a PR, but couldn't because of this issue). 

When I wanted to test the new feature, i switched the dependency to my fork. And i got this weird error above. I thought it was related to my changes. So I reverted my changes. The error persisted. I changed the dependency from my fork to your official repo. And tried the following:
- master branch
- v0.1.x branch
- all 4 commits for `appender` related to version 0.2.2 under v0.1.x branch
None of them worked...

**Where is the actual code for ``tracing-appender = "0.2.2"`` ?

**Comment by hawkw** [comment]
Yeah, I believe the issue here is that all the crates in this repository have `path` dependencies on each other, to facilitate testing of cross-crate changes. So, when taking a Git dependency on one `tracing` crate, it's necessary to patch any other `tracing` crates that are in use to also have a Git dependency on the same branch/SHA. Otherwise, the crate that's depended on from Git will depend on incompatible versions of its dependencies.

**Comment by ozgunozerk** [comment]
Please enlighten me more on this:
in my cargo.toml:
```
tracing = "0.1.31"
tracing-appender = "0.2.2"
tracing-bunyan-formatter = "0.3.2"
tracing-subscriber = "0.3.11"
```

using the `0.2.2` version of `tracing-appender`, I don't have any issues. And when I point to `git` (tried all the revisions for 0.2.2), as 
`tracing-appender = { git = "https://github.com/tokio-rs/tracing", branch = "v0.1.x" }`, why do I have to patch the other `tracing` libraries as well?

Is the released version in `crates.io` is patched? If not, how does the unpatched code from `crates.io` works fine? 
There ought to be a difference between `crates.io` and `git` code in here? Am I wrong?

**Comment by ozgunozerk** [comment]
I've made progress on this bug. And I still think there is an inconsistency.
To reproduce the bug, clone the `tracing` repo (we will use it in the repo below).
To make sure we are in the same version with crates.io, checkout to `b9da5441ea0de983e6ad9c2e8c342938779f3e95` in the tracing repo (this is 0.3.11 for subscriber).
also, clone this: https://github.com/ozgunozerk/tracing-bug-minimal

When you try to run it, it will fail. 
Now, to make it work: go into your `tracing` repo clone. 
`tracing/tracing-appender/Cargo.toml` -> comment line 29 (`path = "../tracing-subscriber"`)

It will work...

P.S. in line 30, the version for subscriber is `0.3`, it will evaulate to `0.3.11` afaik. But to make sure, I also set it to `0.3.11`. Results were the same.

No changes for other crates are necessary. The clone repo will work fine for referencing `tracing, tracing-core`, and etc.
Seems like `tracing-subscriber` is inconsistent with the crates.io.

I tried to debug further using the content of `tracing-subscriber` located in the `.cargo` folder, but could not succeed.

**Comment by davidbarsky** [comment]
> Does the released version of tracing-appender in crates.io include patched version of other tracing crates? If not, how does the unpatched code from crates.io works fine?
> There ought to be a difference between crates.io and git code in here? Am I wrong?

The version released on crates.io does not include patched versions of tracing crates. When publishing, Cargo is able to select between _either_ the path dependency or the version released on crates.io. There is no difference between the version on crates.io versus GitHub, barring unreleased changes. To address any issues, I've pushed a small change to https://github.com/ozgunozerk/tracing-keep-last-n-logs/tree/v0.1.x and opened PR on https://github.com/ozgunozerk/tracing-bug-minimal/pull/1 to show how to patch dependencies.
